src/prototree/base.py

```diff
@@ -0,0 +1,153 @@
+
+
+class ProtoBase(nn.Module):
+    def __init__(
+        self,
+        num_prototypes: int,
+        prototype_shape: tuple[int, int, int],
+        backbone: torch.nn.Module,
+        add_on_layers: Optional[Union[nn.Module, Literal["default"]]] = "default",
+    ):
+        """
+        :param prototype_shape: shape of the prototypes. (channels, height, width)
+        :param backbone: usually a pretrained network that extracts features from the input images
+        :param add_on_layers: used to connect the feature net with the prototypes.
+        """
+        super().__init__()
+
+        # TODO: The parameters in this hardcoded initialization seem to (very, very) roughly correspond to a random
+        #  average-looking latent patch (which is a good start point for the prototypes), but it would be nice if we had
+        #  a more principled way of choosing the initialization.
+        # NOTE: The paper means std=0.1 when it says N(0.5, 0.1), not var=0.1.
+        self.proto_layer = L2Conv2D(
+            num_prototypes, *prototype_shape, initial_mean=0.5, initial_std=0.1
+        )
+
+        if isinstance(add_on_layers, nn.Module):
+            self.add_on = add_on_layers
+        elif add_on_layers is None:
+            self.add_on = nn.Identity()
+        elif add_on_layers == "default":
+            self.add_on = default_add_on_layers(backbone, prototype_shape[0])
+        self.backbone = backbone
+
+    def extract_features(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Applies the feature net and add_on layers to the input. The output
+        has the shape (batch_size, num_channels, height, width), where num_channels is the
+        number of channels of the prototypes.
+        """
+        x = self.backbone(x)
+        x = self.add_on(x)
+        return x
+
+    def patches(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Get the patches for a given input tensor. This is the same as extract_features, except the output is unfolded to
+        be (batch_size, d, n_patches_w, n_patches_h, w_proto, h_proto).
+        """
+        features = self.extract_features(x)
+        return self._features_to_patches(features)
+
+    def distances(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Computes the minimal distances between the prototypes and the input.
+        The output has the shape (batch_size, num_prototypes, n_patches_w, n_patches_h)
+        """
+        x = self.extract_features(x)
+        return self.proto_layer(x)
+
+    def patches_and_dists(self, x: torch.Tensor):
+        features = self.extract_features(x)
+        return self._features_to_patches(features), self.proto_layer(features)
+
+    def _features_to_patches(self, features: torch.Tensor) -> torch.Tensor:
+        w_proto, h_proto = self.prototype_shape[:2]
+        return features.unfold(2, w_proto, 1).unfold(3, h_proto, 1)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
```

Comment:
pls add a description of the output shape

Reply:
Done